src/aws-cpp-sdk-core/source/http/crt/CRTHttpClient.cpp

```diff
@@ -0,0 +1,639 @@
+    {
+        if (!m_client.ContinueRequest(m_currentRequest) || !m_client.IsRequestProcessingEnabled())
+        {
+            return false;
+        }
+
+        bool isAwsChunked = m_currentRequest.HasHeader(Aws::Http::CONTENT_ENCODING_HEADER) &&
+            m_currentRequest.GetHeaderValue(Aws::Http::CONTENT_ENCODING_HEADER) == Aws::Http::AWS_CHUNKED_VALUE;
+
+        size_t amountToRead = buffer.capacity - buffer.len;
+        uint8_t* originalBufferPos = buffer.buffer;
+
+        // aws-chunk = hex(chunk-size) + CRLF + chunk-data + CRLF
+        // Needs to reserve bytes of sizeof(hex(chunk-size)) + sizeof(CRLF) + sizeof(CRLF)
+        if (isAwsChunked)
+        {
+            Aws::String amountToReadHexString = Aws::Utils::StringUtils::ToHexString(amountToRead);
+            amountToRead -= (amountToReadHexString.size() + 4);
+        }        
+
+        // initial check to see if we should avoid reading for the moment.
+        if (!m_rateLimiter || (m_rateLimiter && m_rateLimiter->ApplyCost(0) == std::chrono::milliseconds(0))) {
+            size_t currentPos = buffer.len;
+
+            // now do the read. We may over read by an IO buffer size, but it's fine. The throttle will still
+            // kick-in in plenty of time.
+            bool retValue = Crt::Io::StdIOStreamInputStream::ReadImpl(buffer);
+            size_t newPos = buffer.len;
+            AWS_ASSERT(newPos >= currentPos && !"the buffer length should not have decreased in value.");
+
+            if (retValue && m_isStreaming)
+            {
+                Crt::Io::StreamStatus streamStatus;
+                GetStatus(streamStatus);
+
+                if (newPos == currentPos && !streamStatus.is_end_of_stream && streamStatus.is_valid)
+                {
+                    return true;
+                }
+            }
+            
+            size_t amountRead = newPos - currentPos;
+
+            if (isAwsChunked)
+            {
+                // if we have a chunk to wrap, wrap it, be sure to update the running checksum.
+                if (amountRead > 0)
+                {
+                    if (m_currentRequest.GetRequestHash().second != nullptr)
+                    {
+                        m_currentRequest.GetRequestHash().second->Update(reinterpret_cast<unsigned char*>(originalBufferPos), amountRead);
+                    }
+
+                    Aws::String hex = Aws::Utils::StringUtils::ToHexString(amountRead);
+                    // this is safe because of the isAwsChunked branch above.
+                    // I don't see a aws_byte_buf equivalent of memmove. This is lifted from the curl implementation.
+                    memmove(originalBufferPos + hex.size() + 2, originalBufferPos, amountRead);
+                    memmove(originalBufferPos + hex.size() + 2 + amountRead, "\r\n", 2);
+                    memmove(originalBufferPos, hex.c_str(), hex.size());
+                    memmove(originalBufferPos + hex.size(), "\r\n", 2);

```

Comment:
debatable: this aws-chunked stuff is a lot of copy/pasted finicky code. Consider a utility function that both the CurlHttpClient and CRTHttpClient can share